src/ast.ml

```diff
@@ -110,13 +110,14 @@ type type_ = (* ?type_ *)
   | InductiveType of string * type_ list
   | PredType of string list * type_ list * int option * inductiveness (* if None, not necessarily precise; if Some n, precise with n input parameters *)
   | PureFuncType of type_ * type_  (* Curried *)
-  | ObjType of string
+  | ObjType of string * type_ list (* type arguments *)
   | ArrayType of type_
   | StaticArrayType of type_ * int (* for array declarations in C *)
   | BoxIdType (* box type, for shared boxes *)
   | HandleIdType (* handle type, for shared boxes *)
   | AnyType (* supertype of all inductive datatypes; useful in combination with predicate families *)
-  | TypeParam of string (* a reference to a type parameter declared in the enclosing datatype/function/predicate *)
+  | RealTypeParam of string (* a reference to a type parameter declared in the enclosing Real code *)
+  | GhostTypeParam of string (* a reference to a type parameter declared in the ghost code *)

```

Comment:
The names your propose seem very well-chosen, but in the interest of minimizing conflicts with concurrent PRs being worked on (I am working on several myself, and I know of at least one external party in the middle of a significant effort at this time), I would suggest to leave the name of TypeParam unchanged.

Reply:
Actually, on second thought, it is very little work for the concurrent authors to replace TypeParam by GhostTypeParam in their code; also, having TypeParam and RealTypeParam would make it harder to search just for TypeParam. So, you can leave the names as you propose: RealTypeParam and GhostTypeParam.